### PR TITLE
 Delete blob files only after destroy column handle

### PR DESCRIPTION
**Pull request by Connor1996**
As Rocksdb described, `DropColumnFamilies()` only records the drop of the column family specified by ColumnFamilyHandle. The actual data is not deleted until the client calls `delete column_family`, namely `DestroyColumnFamilyHandle()`. We can still continue using the column family if we have outstanding ColumnFamilyHandle pointer.	

now there are some problems:

- after dropping column family, the blob files of it may be deleted before column family handle is destroyed.
- When opening DB without some existing column families(not supported yet), the blob file of these will be regarded as obsolete files and deleted. 
- After destroyed column handle(not drop the column, just not need to read/write it anymore), we need to remove inner blob storage and evict related cache.

### CODE REVIEW

File: src/version_set.cc [hunk 1]
```diff
@@ -307,16 +323,15 @@ void VersionSet::GetObsoleteFiles(ObsoleteFiles* obsolete_files, SequenceNumber
...
           column_families_.erase(it);
-          obsolete_columns_.erase(cf_id);
         }
       } else {
         fprintf(stderr, "column %u not found when deleting obsolete file%" PRIu64 "\n", 
           cf_id, file_number);
         abort();        
       }
-      auto now = tuple_it++;
+      auto now = tuple_it++; // is that okay?
```
Comment (yiwu-arbug) [inline src/version_set.cc]: not sure if it is okay either. change it to `tuple_it = obsolete_files->blob_files.erase(tuple_it)` may be better.

**Comment by Connor1996** [comment]
Yes, it was late yesterday so I just left some unfinished comments and test to do.  Working on that~

File: src/version_set.cc (hunk 1 above)
Comment (Connor1996) [reply src/version_set.cc]: from https://en.cppreference.com/w/cpp/container/list/splice
`No elements are copied or moved, only the internal pointers of the list nodes are re-pointed.`
Seems it is fine.

File: src/db_impl.cc [hunk 2]
```diff
@@ -286,25 +286,22 @@ Status TitanDBImpl::CreateColumnFamilies(
...
-  // As rocksdb described, `DropColumnFamilies()` only records the drop of the column family specified by ColumnFamilyHandle.
-  // The actual data is not deleted until the client calls `delete column_family`, namely `DestroyColumnFamilyHandle()`.
-  // We can still continue using the column family if we have outstanding ColumnFamilyHandle pointer.
-  // So we should delete blob files in `DestroyColumnFamilyHandle()` but not here.
-  Status s = db_impl_->DropColumnFamilies(handles);
   if (s.ok()) {
-    SequenceNumber obsolete_sequence = db_impl_->GetLatestSequenceNumber();
-    vset_->DropColumnFamilies(column_families, obsolete_sequence);
+    MutexLock l(&mutex_);
+    vset_->DestroyColumnFamily(cf_id);
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/db_impl.cc]: Yes, it does not delete the CF, `vset_->DestroyColumnFamily` just makes some marks.

**Comment by Connor1996** [comment]
All works have done, PTAL @yiwu-arbug

File: src/db_impl.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply src/db_impl.cc]: I mean, for example:
* open db, and store CF handle in `cf_handle`
* call `DestroyColumnFamilyHandle(cf_handle)`.

In this case we should not mark the CF as destroyed. It simply means the user don't want to read from this CF any more. On db reopen this CF is still a live CF for normal read and write.

[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply src/db_impl.cc]: I mean, if the CF is not in `obsolete_columns_`, we should not mark it as destroyed here.

[the remark this answers is not in the document]
Comment (Connor1996) [reply src/db_impl.cc]: We should. When the column family handle is destroyed, we can remove the corresponding blob storage in memory to save memory and evict cache.

[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply src/db_impl.cc]: Add some comments here? the word "destroy" may sound like we are deleting blob files, while we may not.

File: src/blob_storage.h [hunk 3]
```diff
@@ -79,6 +100,12 @@ class BlobStorage {
   std::shared_ptr<BlobFileCache> file_cache_;
 
   std::vector<GCScore> gc_score_;
+
+  std::list<std::pair<uint64_t, SequenceNumber>> obsolete_files_;
+  // It is marked when the column family handle is destroyed, indicating this
+  // structure can be removed after all the blob files of this column family are
```
Comment (yiwu-arbug) [inline src/blob_storage.h]: the comment sounds misleading. It only means the in-memory data structure can be destroyed. Physical files may still be kept.

**Pull request merged**